tokio/src/task/task_set.rs

```diff
@@ -0,0 +1,222 @@
+use std::fmt;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+use crate::runtime::Handle;
+use crate::task::{JoinError, JoinHandle, LocalSet};
+use crate::util::IdleNotifiedSet;
+
+/// A collection of tasks spawned on a Tokio runtime. A `TaskSet` is not ordered, and the tasks
+/// will be returned in the order they completed.
+///
+/// All of the tasks must have the same return type `T`.
+///
+/// When the `TaskSet` is dropped, all tasks in the `TaskSet` are immediately aborted.
+///
+/// # Examples
+///
+/// Spawn multiple tasks and wait for them.
+///
+/// ```
+/// use tokio::task::TaskSet;
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let mut set = TaskSet::new();
+///
+///     for i in 0..10 {
+///         set.spawn(async move { i });
+///     }
+///
+///     let mut seen = [false; 10];
+///     while let Some(res) = set.join_one().await.unwrap() {
+///         seen[res] = true;
+///     }
+///
+///     for i in 0..10 {
+///         assert!(seen[i]);
+///     }
+/// }
+/// ```
+pub struct TaskSet<T> {
```

Comment:
IIRC we discussed renaming this to `JoinSet`? What was the conclusion there? If I remember, the question was whether `TaskSet` would be vague compared to `LocalSet` which is also a "task set" (we have a lot of sets...).

Reply:
I prefer TaskSet because it goes along with the broader scoping/cancellation features.